File: sdk/python/alp_sdk/code_transform.py
from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Dict, Any, Optional

class CodeTransformResult:
    __test__ = False

    def __init__(
        self,
        transform_id: str,
        transform_type: str,
        target_file: str,
        original_code: str,
        transformed_code: str,
        diff_preview: str,
        status: str = "applied",
        applied_at: Optional[str] = None,
    ):
        self.id = transform_id
        self.transform_type = transform_type
        self.target_file = target_file
        self.original_code = original_code
        self.transformed_code = transformed_code
        self.diff_preview = diff_preview
        self.status = status
        self.applied_at = applied_at or datetime.now(timezone.utc).isoformat()

class CodeTransformConfig:
    __test__ = False

    def __init__(
        self,
        transform_id: str,
        transform_type: str,
        target_file: str,
        target_symbol: Optional[str] = None,
        new_symbol: Optional[str] = None,
        diff_preview: Optional[str] = None,
        status: str = "pending",
        description: Optional[str] = None,
    ):
        self.id = transform_id
        self.transform_type = transform_type
        self.target_file = target_file
        self.target_symbol = target_symbol
        self.new_symbol = new_symbol
        self.diff_preview = diff_preview
        self.status = status
        self.description = description

class CodeTransformEngine:
    def __init__(self):
        self.transforms: Dict[str, CodeTransformResult] = {}

    def apply_transform(
        self,
        transform_id: str,
        transform_type: str,
        target_file: str,
        source_code: str,
        target_symbol: Optional[str] = None,
        new_symbol: Optional[str] = None,
    ) -> CodeTransformResult:
        transformed_code = source_code

        if transform_type == "rename_symbol" and target_symbol and new_symbol:
            pattern = rf"\b{re.escape(target_symbol)}\b"
            transformed_code = re.sub(pattern, new_symbol, source_code)
        elif transform_type == "add_log_guard":
            indented = "\n".join("  " + line for line in source_code.split("\n"))
            transformed_code = f"// [ALP Guarded Execution]\ntry {{\n{indented}\n}} catch (err) {{\n  console.error('[ALP Guard] Error:', err);\n}}"
        elif transform_type == "extract_function":
            func_name = new_symbol or "extracted_helper"
            transformed_code = f"def {func_name}():\n    pass # Extracted logic\n\n{source_code}"
        elif transform_type == "migration_rewrite":
            transformed_code = re.sub(r"var\s+", "let ", source_code)

        orig_lines = len(source_code.split("\n"))
        trans_lines = len(transformed_code.split("\n"))
        diff_preview = f"--- {target_file}\n+++ {target_file} (transformed)\n@@ -1,{orig_lines} +1,{trans_lines} @@\n{transformed_code[:150]}..."

        result = CodeTransformResult(
            transform_id=transform_id,
            transform_type=transform_type,
            target_file=target_file,
            original_code=source_code,
            transformed_code=transformed_code,
            diff_preview=diff_preview,
            status="applied",
        )

        self.transforms[transform_id] = result
        return result

    def revert_transform(self, transform_id: str) -> Optional[CodeTransformResult]:
        transform = self.transforms.get(transform_id)
        if not transform:
            return None
        transform.status = "reverted"
        transform.transformed_code = transform.original_code
        return transform

    def get_transform(self, transform_id: str) -> Optional[CodeTransformResult]:
        return self.transforms.get(transform_id)

File: parser/src/code-transform.ts
export type TransformType = 'rename_symbol' | 'extract_function' | 'inline_variable' | 'add_log_guard' | 'migration_rewrite';
export type TransformStatus = 'pending' | 'applied' | 'reverted';

export interface CodeTransformResult {
  id: string;
  transformType: TransformType;
  targetFile: string;
  originalCode: string;
  transformedCode: string;
  diffPreview: string;
  status: TransformStatus;
  appliedAt?: string;
}

export interface CodeTransformConfig {
  id: string;
  transformType: TransformType;
  targetFile: string;
  targetSymbol?: string;
  newSymbol?: string;
  diffPreview?: string;
  status?: TransformStatus;
  description?: string;
}

export class CodeTransformEngine {
  private transforms: Map<string, CodeTransformResult> = new Map();

  public applyTransform(
    id: string,
    transformType: TransformType,
    targetFile: string,
    sourceCode: string,
    targetSymbol?: string,
    newSymbol?: string
  ): CodeTransformResult {
    let transformedCode = sourceCode;

    switch (transformType) {
      case 'rename_symbol':
        if (targetSymbol && newSymbol) {
          const regex = new RegExp(`\\b${targetSymbol}\\b`, 'g');
          transformedCode = sourceCode.replace(regex, newSymbol);
        }
        break;
      case 'add_log_guard':
        transformedCode = `// [ALP Guarded Execution]\ntry {\n${sourceCode.replace(/^/gm, '  ')}\n} catch (err) {\n  console.error("[ALP Guard] Execution error:", err);\n}`;
        break;
      case 'extract_function':
        const extractedName = newSymbol || 'extractedHelper';
        transformedCode = `function ${extractedName}() {\n  // Extracted logic block\n}\n\n${sourceCode}`;
        break;
      case 'inline_variable':
        if (targetSymbol && newSymbol) {
          const regex = new RegExp(`const\\s+${targetSymbol}\\s*=\\s*[^;]+;\\s*`, 'g');
          transformedCode = sourceCode.replace(regex, '');
          transformedCode = transformedCode.replace(new RegExp(`\\b${targetSymbol}\\b`, 'g'), newSymbol);
        }
        break;
      case 'migration_rewrite':
        transformedCode = sourceCode.replace(/var\s+/g, 'let ');
        break;
    }

    const diffPreview = `--- ${targetFile}\n+++ ${targetFile} (transformed)\n@@ -1,${sourceCode.split('\n').length} +1,${transformedCode.split('\n').length} @@\n${transformedCode.slice(0, 150)}...`;

    const result: CodeTransformResult = {
      id,
      transformType,
      targetFile,
      originalCode: sourceCode,
      transformedCode,
      diffPreview,
      status: 'applied',
      appliedAt: new Date().toISOString(),
    };

    this.transforms.set(id, result);
    return result;
  }

  public revertTransform(id: string): CodeTransformResult | undefined {
    const transform = this.transforms.get(id);
    if (!transform) return undefined;

    transform.status = 'reverted';
    transform.transformedCode = transform.originalCode;
    return transform;
  }

  public getTransform(id: string): CodeTransformResult | undefined {
    return this.transforms.get(id);
  }
}
